Source/Scene/modelMaterialsCommon.js

```diff
@@ -0,0 +1,771 @@
+                        fragmentLightingBlock += '    float attenuation = 1.0 / (u_' + lightBaseName + 'Attenuation.x + ';
+                        fragmentLightingBlock += '(u_' + lightBaseName + 'Attenuation.y * range) + ';
+                        fragmentLightingBlock += '(u_' + lightBaseName + 'Attenuation.z * range * range));\n';
+                    }
+                    else {
+                        fragmentLightingBlock += '    float attenuation = 1.0;\n';
+                    }
+
+                    if (lightType === 'spot') {
+                        fragmentLightingBlock += '    float spotDot = dot(-l, normalize(' + varyingDirectionName + '));\n';
+                        fragmentLightingBlock += '    if (spotDot < cos(u_' + lightBaseName + 'FallOff.x * 0.5))\n';
+                        fragmentLightingBlock += '    {\n';
+                        fragmentLightingBlock += '      attenuation = 0.0;\n';
+                        fragmentLightingBlock += '    }\n';
+                        fragmentLightingBlock += '    else\n';
+                        fragmentLightingBlock += '    {\n';
+                        fragmentLightingBlock += '        attenuation *= max(0.0, pow(spotDot, u_' + lightBaseName + 'FallOff.y));\n';
+                    }
+
+                    fragmentLightingBlock += '    diffuseLight += ' + lightColorName + '* max(dot(normal,l), 0.) * attenuation;\n';
+
+                    if (hasSpecular) {
+                        if (lightingModel === 'BLINN') {
+                            fragmentLightingBlock += '    vec3 h = normalize(l + viewDir);\n';
+                            fragmentLightingBlock += '    float specularIntensity = max(0., pow(max(dot(normal, h), 0.), u_shininess)) * attenuation;\n';
+                        }
+                        else { // PHONG
+                            fragmentLightingBlock += '    vec3 reflectDir = reflect(-l, normal);\n';
+                            fragmentLightingBlock += '    float specularIntensity = max(0., pow(max(dot(reflectDir, viewDir), 0.), u_shininess)) * attenuation;\n';
+                        }
+                        fragmentLightingBlock += '    specularLight += ' + lightColorName + ' * specularIntensity;\n';
+                    }
+
+                    if (lightType === 'spot') {
+                        fragmentLightingBlock += '    }\n';
+                    }
+                }
+                fragmentLightingBlock += '  }\n';
+            }
+        }
+
+        if (!hasAmbientLights) {
+            // Add an ambient light if we don't have one
+            fragmentLightingBlock += '  ambientLight += vec3(0.1, 0.1, 0.1);\n';
+        }
+
+        if (!hasNonAmbientLights) {
+            fragmentLightingBlock += '  vec3 l = normalize(czm_sunDirectionEC);\n';
+            fragmentLightingBlock += '  diffuseLight += vec3(1.0, 1.0, 1.0) * max(dot(normal,l), 0.);\n';
```

Comment:
I got shader compilation errors here when the material is constant. I ended up doing this, but you could do it differently:

`if (!hasNonAmbientLights && (lightingModel !== 'CONSTANT')) {`